### DISCUSSION TITLE
Use phpspec/prophecy as a mocking Framework instead of PhpUnit

### DISCUSSION THREAD

**Discussion by xserrat** [pheature-flags/pheature-flags#154] [Ideas]
Hello!

I was thinking of using [phpspec/prophecy](https://github.com/phpspec/prophecy) instead of PhpUnit as a mocking Framework.

### Benefits
* ~~You can mock classes defined as `final`~~ -> **EDITED**: It's not possible. See next comment.
* If you use PhpStorm as IDE, the refactor tool to rename a method of a class will also rename the method of the mocked class because you define the expectation calling the method instead of passing the name of the method as string.

### Drawbacks
* It's not an standard in a PHP library
* We need to install an extra vendor [phpspec/prophecy-phpunit](https://github.com/phpspec/prophecy-phpunit) in addition to the main vendor [phpspec/prophecy](https://github.com/phpspec/prophecy) due to this deprecation warning:
> PHPUnit\Framework\TestCase::prophesize() is deprecated and will be removed in PHPUnit 10. Please use the trait provided by phpspec/prophecy-phpunit.

### Reputation:
* PhpUnit: ⭐️ 17.5k, Forks: 2k
* Prophecy: ⭐️ 8k, Forks: 209

Here we have an example using both libraries:

**Using Prophesize**:

```php
class RemoveStrategyFactoryTest extends TestCase
{
    use ProphecyTrait;

    public function testItShouldCreateARemoveStrategyFromInvokable(): void
    {
        $expected = $this->prophesize(FeatureRepository::class)->reveal();

        $mockedContainer = $this->prophesize(ContainerInterface::class);
        $mockedContainer
            ->get(FeatureRepository::class)
            ->shouldBeCalledOnce()
            ->willReturn($expected);

        $factory = new RemoveStrategyFactory();
        $actual = $factory->__invoke($mockedContainer->reveal());

        self::assertInstanceOf(RemoveStrategy::class, $actual);
    }
}
```

**Using PhpUnit**:

```php
class RemoveStrategyFactoryTest extends TestCase
{
    public function testItShouldCreateARemoveStrategyFromInvokable(): void
    {
        $mockedContainer = $this->createConfiguredMock(ContainerInterface::class, [
            'get' => $this->createMock(FeatureRepository::class)
        ]);

        $factory = new RemoveStrategyFactory();
        $actual = $factory->__invoke($mockedContainer);

        self::assertInstanceOf(RemoveStrategy::class, $actual);
    }
}
```

**Comment by kpicaza** [comment]
It has some trade-offs. As commented is not the standard testing library in PHP and it adds two dev dependencies. But on the other hand, the ability to mock final classes will help us a lot in testing the other libraries.

I think that Procecy can add value to our project making it simpler to do the unit tests in many cases. Go for it :rocket::rocket::rocket:

**Comment by xserrat** [comment]
I was making the test of the `PostFeatureFactory` to check how it'd be but it's not possible the mocking of `final` classes... It's something requested in [this](https://github.com/phpspec/prophecy/issues/197) issue but it's no way with `Prophecy`. The suggestions they recommend are:
* Add an interface -> I think it doesn't make sense if we're not coupled to any I/O system or third party library...
* Use the [bypass-finals](https://github.com/dg/bypass-finals) library that modifies your PHP final classes removing the `final` declaration` to enable the mocking. 
* Instantiate the `final` class even if we're testing more logic than needed i.e: the [Symfony ExceptionEvent](https://github.com/api-platform/core/blob/main/tests/EventListener/ExceptionListenerTest.php#L42) instantiated in the unit test of the `ExceptionListenerTest` of ApiPlatform.

The problem of the last suggestion is that if the constructor changes, we'll have to update all tests that are instantiating the class. If it was a problem, we could extract into a single class for testing purposes to centralise these changes to a single point.

Anyway, if Prophecy makes us a simpler tests, it would be a good benefit to use it!

**Reply by kpicaza** [reply]
I don't see a problem with using our final classes in test cases, most cases are value objects that are not too hard to instantiate, and every glue class has its own interface(if not we can add it).

In projects like laminas, they had opted to remove prophecy in some cases because of complexity https://github.com/laminas/laminas-component-installer/pull/25.

I'm not sure about this, here is a post by @TomasVotruba explaining how to use the [bypass-finals](https://github.com/dg/bypass-finals) package with PHPUnit framework https://tomasvotruba.com/blog/2019/03/28/how-to-mock-final-classes-in-phpunit/

**Reply by TomasVotruba** [reply]
Thanks for sharing. The `final` is not really needed for mocking :+1: 
Adding interface for mocking only confuses contributors.

Another point of view on writing clear PHP code without mocks: https://tomasvotruba.com/blog/2018/06/11/how-to-turn-mocks-from-nightmare-to-solid-kiss-tests/

**Reply by kpicaza** [reply]
Thanks!!!! very interesting, I like the idea;-D

**Reply by xserrat** [reply]
Thank you so much @TomasVotruba for your contribution! It's a great approach to make tests simpler to understand without third party libraries 👌🏽

**Reply by kpicaza** [reply]
I've tried the anon class approach testing the `ToggleStrategyFactoryPass` class https://github.com/pheature-flags/pheature-flags/pull/180/files#diff-7d850b8461d5bc3608fcc46bca023e0e5a858ec2ac291794005a60b6bf3f610fR42. I'm not sure if it is correctly applied.

**Reply by TomasVotruba** [reply]
Using anonymous classes there seems too magic to me. I know what it is, but I still struggle with reading that code.

When working with standard packages like Symfony, I would use the same syntax that end developers doo. YAML or PHP and load the config. It makes maintaining much easier, as developers don't have to learn new opinionated syntax, but they use what they use in every Symfony app.

### Have a factory class - this can be re-usable in any other place that is using config:

```php
use Psr\Container\ContainerInterface;
use Symfony\Component\Config\FileLocator;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;

final class ContainerFactory
{
    public function create(string $configFile, CompilerPassInterface $compilerPass): ContainerInterface
    {
        $containerBuilder = new ContainerBuilder();
        $containerBuilder->addCompilerPass($compilerPass);

        $phpFileLoader = new YamlFileLoader($containerBuilder, new FileLocator());
        $phpFileLoader->load($configFile);

        $containerBuilder->compile();

        return $containerBuilder;
    }
}
```

### Then use in your code:

```php
$containerFactory = new ContainerFactory();
$contaienrBuilder = $containerFactory->create(__DIR__ . '/config/config.yaml', $yourCompilerPass);
```

with `config.yaml`:

```yaml
extension_name:
    key: value
```

**Reply by kpicaza** [reply]
Thanks :top:. Totally agree. We really appreciate your help, and we are very happy with all that we are learning;-D.

**Reply by TomasVotruba** [reply]
Thank you :)

Don't forget to take me critically and question my suggestion. It's just one man's view on coding. And I'm extremely lazy man :D

**Reply by kpicaza** [reply]
Sure ;-D, in this case, it has all sense, because we have to apply the same test type to at least 4 other compiler passes. Also, we haven't too much experience in testing Symfony bundles. Thanks really,
